### DISCUSSION TITLE
Pass child local to parent local?

### DISCUSSION THREAD

**Discussion by tracphil** [gruntwork-io/knowledge-base#734] [Help] [answered]
The subject is the best way to describe what I am trying to do.

I am trying to merge a local that appears down in the filesystem into a local that is in `_common/common.hcl`.

The error that I am getting is:

```
aws_account_alias [REASON: Can't evaluate expression at (filesystem_path)/_common/common.hcl:4,22-66 because local reference 'environment' is not evaluated.
```

This is how I have things setup.

`_common/common.hcl`

```hcl
locals {
  aws_account_alias= "networking-${local.environment}-01"
}
```

`environments/account-dev/account.hcl`

```hcl
locals {
  account_id = "12345678901"
  environment = "dev"
}
```

`environments/networking-dev/us-east-1/terragrunt.hcl`

```hcl
locals {
  aws_region = "us-east-1"
}
include "root" {
  path = find_in_parent_folders()
}
```

`terragrunt.hcl`

This is the relevant part of `terragrunt.hcl`:

```hcl
locals {
  account_vars = read_terragrunt_config(find_in_parent_folders("account.hcl"))
  common_vars = read_terragrunt_config(find_in_parent_folders("_common/common.hcl"))
}
```

Are locals evaluated in the order they are included?

Thanks for any help you can provide.

---

<ins datetime="2023-06-08T19:41:19Z">
  <p><a href="https://support.gruntwork.io/hc/requests/110244">Tracked in ticket #110244</a></p>
</ins>

**Comment by tracphil** [comment]
I should add. I am testing this with `terragrunt render-json`. I am new to Terragrunt. Is there another way to test that your module will render?

**Comment by denis256** [comment, marked as answer]
Hello,
configuration can be extracted to separated HCL file and loaded in parent file through `read_terragrunt_config()` using functions to get paths `get_terragrunt_dir()` or `path_relative_to_include()`

I also prepared simplified example in https://github.com/denis256/terragrunt-tests/tree/master/discussion-734

References:

https://terragrunt.gruntwork.io/docs/reference/built-in-functions/#read_terragrunt_config
https://terragrunt.gruntwork.io/docs/reference/built-in-functions/#get_terragrunt_dir
https://terragrunt.gruntwork.io/docs/reference/built-in-functions/#path_relative_to_include

**Reply by tracphil** [reply]
Thank you so much!

**Reply by scr-oath** [reply]
This seems fine for the "parent" but I'm feeling like I'm repeating myself quite often with the individual "_env" includes…

I'd like to conceptually have a merge(common, account, region) as a common vars - to mimic the places that differentiate - and end up with an easy to use common_vars, and to use this in the root terragrunt.hcl as well as my various _env things like _env/app1.hcl _env/app2.hcl, which are read in … a la the examples in https://terragrunt.gruntwork.io/docs/features/keep-your-terragrunt-architecture-dry/

```
include "root" {
  path = find_in_parent_folders()
}

include "env" {
  path = "${get_terragrunt_dir()}/../../_env/app1.hcl"
}
```

If I try to make a common.hcl at the top and do

```
locals {
  common_vars = read_terragrunt_config(find_in_parent_folders("common.hcl"))
  account-id = local.common_vars.locals.common.account-id
}
```

where common.hcl looks like 
```
locals {
  account_vars = read_terragrunt_config(find_in_parent_folders("account.hcl"))
  region_vars = read_terragrunt_config(find_in_parent_folders("region.hcl"))
  common_vars = {
   owner = "myteam@myorg.com"
   repo = "path-to-git-repo"
  }
  common = merge(
    local.common_vars,
    local.account_vars,
    local.region_vars,
  )
}
```

But… it seems that, when reading the common.hcl the dir for the `find_in_parent_folders` shifts from the "child" to the common.hcl - so it cannot find the child-specific vars to be helpful here and be DRY…

Can you help with a good mechanism for reducing repetition for common map of variables merged from least specific to most specific and with info pertaining to the child?
